### PR TITLE
fix: move compression.go into confighttp.go

### PR DESCRIPTION
**Pull request by hyunuk**
**Description:**

This PR moved every code in `internal/middleware/compression.go` to `config/confighttp.go` and tests as well.
It also changed functions unexported except `CompressionType`.

_Note: Functions in `compression.go` was iniatially created for `otlphttpexporter`, but it was changed to the `confighttp` in this PR(see #4441 and [the code](https://github.com/open-telemetry/opentelemetry-collector/pull/4441/files#diff-a13e181e395f2cb03704cecf9c08f081decf273f1042b7244ad732b01637ce24L93-L99))._

**Link to tracking Issue:**
#4627

### CODE REVIEW

**Review by bogdandrutu — COMMENTED** [review]
Not sure if it is ok to make `configgrpc` depend on `confighttp`. Can we maybe consider a `configcompression`? Idea being other protocols (e.g. plain UDP) can use these constants as well.

Another option is `confignet`? Personally I prefer `configcompression`, any thoughts @tigrannajaryan @open-telemetry/collector-approvers ?

File: config/configcompression/compressionType.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+type CompressionType string
+
+const (
+	CompressionGzip    CompressionType = "gzip"
+	CompressionZlib    CompressionType = "zlib"
+	CompressionDeflate CompressionType = "deflate"
+	CompressionSnappy  CompressionType = "snappy"
+	CompressionZstd    CompressionType = "zstd"
+	CompressionNone    CompressionType = "none"
+	CompressionEmpty   CompressionType = ""
```
Comment (bogdandrutu) [inline config/configcompression/compressionType.go:28]: Interested about the "empty" value... Should it be a public value for it? What is the use-case for it?

Comment (hyunuk) [reply config/configcompression/compressionType.go:28]: `CompressionEmpty` and `CompressionNone` was used in a different package before but now they are only used in here. I'll make them unexported.

File: config/configgrpc/configgrpc.go [hunk 2]
```diff
@@ -39,6 +39,7 @@ import (
 	"go.opentelemetry.io/collector/client"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/config/configauth"
+	"go.opentelemetry.io/collector/config/configcompression"
```
Comment (jpkrohling) [inline config/configgrpc/configgrpc.go:42]: No strong feelings, but I think I would prefer confignet, if we believe we might need another shared configs than compression.

File: config/configgrpc/configgrpc.go [hunk 3]
```diff
@@ -357,13 +358,13 @@ func (gss *GRPCServerSettings) ToServerOption(host component.Host, settings comp
 }
 
 // getGRPCCompressionName returns compression name registered in grpc.
-func getGRPCCompressionName(compressionType middleware.CompressionType) (string, error) {
+func getGRPCCompressionName(compressionType configcompression.CompressionType) (string, error) {
 	switch compressionType {
-	case middleware.CompressionGzip:
+	case configcompression.CompressionGzip:
```
Comment (jpkrohling) [inline config/configgrpc/configgrpc.go:363]:
```suggestion
	case configcompression.Gzip:
```

If we stick with configcompression, remove "compression" from the property name.

File: CHANGELOG.md [hunk 4]
```diff
@@ -15,6 +15,7 @@
   - The merge functionality is now embedded into `service.NewConfigProvider` (#4637).
 - Move `configtest.LoadConfig` and `configtest.LoadConfigAndValidate` to `servicetest` (#4606)
 - Builder: Remove deprecated `include-core` flag (#4616)
+- Move `compression.go` into `confighttp.go` (#4651)
```
Comment (dmitryax) [inline CHANGELOG.md:18]: It doesn't seem like a breaking change. It's rather an additional public API / enhancement. Also please move it to "Unreleased" section and reword it to indicate what public API is exposed in which package.

File: config/configgrpc/configgrpc.go (hunk 3 above)
Comment (hyunuk) [reply config/configgrpc/configgrpc.go]: done. Thank you

File: CHANGELOG.md (hunk 4 above)
Comment (hyunuk) [reply CHANGELOG.md]: done. Thank you

File: config/configgrpc/configgrpc.go (hunk 2 above)
Comment (bogdandrutu) [reply config/configgrpc/configgrpc.go:42]: I don't mind using the `confignet`, @tigrannajaryan @open-telemetry/collector-approvers please give at least one opinion for `configcompression` vs `confignet`

Comment (codeboten) [reply config/configgrpc/configgrpc.go:42]: I find `configcompression` more meaningful as to what the contents of the package are today. `confignet` doesn't make me think of compression necessarily, and it would be possible for these options to be used outside of network transport (ie for compressing data before saving it to a datastore) in which case it would be confusing for the option to be in `confignet`. It's not a strong preference but just what comes to mind off top my head.

**Review by codeboten — APPROVED** [review]

File: config/confighttp/confighttp.go [hunk 5]
```diff
@@ -325,3 +333,158 @@ func authInterceptor(next http.Handler, authenticate configauth.AuthenticateFunc
 		next.ServeHTTP(w, r.WithContext(ctx))
 	})
 }
+
+type compressRoundTripper struct {
```
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:337]: Can we have this in a separate file in the same dir to avoid super large files?

**Review by bogdandrutu — APPROVED** [review]

File: config/confighttp/confighttp.go (hunk 5 above)
Comment (bogdandrutu) [reply config/confighttp/confighttp.go]: Thank you @hyunuk

**Comment by bogdandrutu** [comment]
please fix lint

File: config/configcompression/compressionType.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+type CompressionType string
+
+const (
+	Gzip    CompressionType = "gzip"
+	Zlib    CompressionType = "zlib"
+	Deflate CompressionType = "deflate"
+	Snappy  CompressionType = "snappy"
+	Zstd    CompressionType = "zstd"
+	None    CompressionType = "none"
```
Comment (bogdandrutu) [inline config/configcompression/compressionType.go:27]: One PR ago you told me that this does not need to be public
